File: cpp/src/euler097.cpp
// Problem 97: Large non-Mersenne prime
// https://projecteuler.net/problem=97
// Find the last ten digits of 28433 × 2^7830457 + 1.

#include <iostream>

long long mod_pow(long long base, long long exp, long long mod) {
    long long result = 1;
    base %= mod;
    while (exp > 0) {
        if (exp % 2 == 1) {
            // Use __int128 to avoid overflow in multiplication
            __int128 temp = (__int128)result * base % mod;
            result = temp;
        }
        __int128 temp_base = (__int128)base * base % mod;
        base = temp_base;
        exp /= 2;
    }
    return result;
}

long long large_non_mersenne_prime() {
    const long long MOD = 10000000000LL; // 10^10
    long long pow2 = mod_pow(2, 7830457, MOD);
    // Use __int128 for the final multiplication to avoid overflow
    __int128 temp = (__int128)28433LL * pow2 % MOD;
    long long result = (temp + 1) % MOD;
    return result;
}

#ifndef UNITTEST_MODE
int main() {
    std::cout << large_non_mersenne_prime() << std::endl;
    return 0;
}
#endif

File: go/euler097.go
// https://projecteuler.net/problem=97

// Large Non-Mersenne Prime

// The first known prime found to exceed one million digits was discovered in 1999, and is a Mersenne prime of the form 2^6972593−1; it contains exactly 2,098,960 digits. Subsequently other Mersenne primes, of the form 2^p−1, have been found which contain more digits.

// However, in 2004 there was found a massive non-Mersenne prime which contains 2,357,207 digits: 28433×2^7830457+1.

// Find the last ten digits of this prime number.

// Answer: 8739992577

// Execution time: ~0ms

// Optimizations: Modular exponentiation using big integers

// Interesting notes: AI-generated solution using Go's big.Int for large number arithmetic

package main

import (
	"fmt"
	"math/big"
)

func largeNonMersennePrime() string {
	mod := big.NewInt(10000000000) // 10^10
	base := big.NewInt(2)
	exp := big.NewInt(7830457)
	pow2 := new(big.Int).Exp(base, exp, mod)
	result := new(big.Int).Mul(big.NewInt(28433), pow2)
	result.Add(result, big.NewInt(1))
	result.Mod(result, mod)
	return result.String()
}

func main() {
	fmt.Println(largeNonMersennePrime())
}